### DISCUSSION TITLE
클라이언트 구매시 OnPurchaseFailed가 실행됩니다

### DISCUSSION THREAD

**Discussion by TaejunPark** [naverz/zepeto-studio-kor#943] [5. Q&A]
안녕하십니까,
클라이언트 구매를 시도해보던 중, 팝업이 떠서 구매버튼을 누르면

OnPurchaseFailed : 
    at PurchaseHandler.OnPurchaseFailed (Assets/GJ/Scripts/PurchaseHandler.js:52)
    at (/Users/taejunpark/Projects/GJ09/builtin/proxy.js:8)

이런식으로 에러 메시지는 뒤에 프린트 안되면서  failed 부분이 실행되는데, 어떻게 하면 구매성공 부분이 실행될까요
함수 이름을 특별히 건드리진 않았습니다
커버이미지는 등록해놓았습니다.

감사합니다.

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, 먼저 친절하고 상세하게 문의를 올려주셔서 감사합니다.
https://studio.zepeto.me/guides/linking-purchasecomplete-event1 
가이드대로 OnPurchaseComplete에도 함수를 세팅하셨다면 동작은 정상적으로 될 것으로 예상합니다. 
의심되는 두 가지 상황은
1. IWP상품이 비소모성인가요? 현재는 에디터에서 비소모성 아이템은 최초 구매 이후 재구매가 불가하여 다음 패치때 개선 예정입니다.
2. 에디터에 로그인된 계정과, 월드를 등록한 계정이 동일한지 확인 부탁 드립니다.
감사합니다.

**Reply by TaejunPark** [reply]
안녕하세요, 항상 친절한 답변 주셔서 감사합니다. 혼자 분해서 삭막하게 썼었는데 반성하게 되네요.ㅠ
개인적으로는 관련된 작업이 끝나서, 해놓은 거만 올려놓고 자주 올 것 같진 않으나(넘 어려워서..),
훌륭한 담당자 분이 있으신 건 계속 기억날 것 같습니다.


1. 두가지를 다 해보긴 하고 새로 만들어서도 해봤는데, 콘솔에 출력 된 내용은 동일한 것 같습니다. 비소모성은 한번 사면 다음에 살땐 구매버튼만 비활성화 됩니다. 혹시 iwp 버전이 안맞나 싶어 패키지를 업데이트 해봤더니, 실행이 잘 안돼서, 월드버전도 새로 업데이트 해서 돌려봤으나 동일한 출력이 나왔습니다.
2. 동일한 것 같습니다

한번 퍼블리시로 올려서 테스트 해봤더니, 스크린샷 2번째 처럼 구매가 완료됐다고 뜨는데,
게임 내 반응(성공/실패 오디오 및 맵 열림/안열림)으로 봐선 OnPurchaseFailed 부분이 실행되는 걸로 보입니다. 

![IMG_0257](https://user-images.githubusercontent.com/3603377/177137796-9fd366a0-27d7-42b0-a27b-a538126e147b.PNG)

![IMG_0258](https://user-images.githubusercontent.com/3603377/177138003-e188d8ff-92b8-473e-b070-d8fa88c4f491.PNG)

관련해서 해결 방안이 있을까요?
감사합니다.

**Reply by Zepeto-TechEvangelist** [reply]
안녕하세요~ 도움을 드리고 싶으니 
 [dl_zepeto_evangelist@naverz-corp.com](mailto:dl_zepeto_evangelist@naverz-corp.com) 메일로
월드 id와 함께 
프로젝트 파일을 전달해 주실 수 있을까요?
감사합니다.

**Reply by TaejunPark** [reply]
보내놓았습니다. 감사합니다~😄😆

**Reply by Zepeto-TechEvangelist** [reply]
넵, 확인 결과 Trigger에서 Purchase Popup을 띄워주는 부분에서 실패가 났었습니다. 

Purchase Block 영역에 진입시 구매 버튼이 보여지고, 
구매 버튼을 눌렀을때 상품이 구매되는 간단한 형태로 변경하니 성공적으로 구매가 되었습니다. (스샷의 아이템 id는 제 로그인 아이디로 등록한 임의의 IWP입니다) 
메일로 간단히 변경한 프로젝트 파일을 전달 드릴 예정이오니 참고하셔서
구현하시면 좋을 것 같습니다 :)

![스크린샷 2022-07-05 오후 5 53 20](https://user-images.githubusercontent.com/100118650/177289913-77a68067-a4ed-49e2-8d42-6877df4f4433.png)

**Reply by TaejunPark** [reply]
ㅎㅎ 전달해주신 내용 잘 참고하였습니다 ㅠ
혹여나 비슷한 느낌으로 틀리시는 분이 있을까봐,, 수다스럽지만 내용을 적어놓습니다🤣😇

1. 제가 작성한 OnPurchaseComplete 중간에 에러나는 부분이 있어서 OnPurchaseFailed가 호출되던거였는데,
로그 찍는 부분을 함수 뒤쪽으로 밀어놨더니 OnPurchaseComplete에 진입했던 걸 모르고 OnPurchaseFailed만 호출되는 줄 알았었습니다

2. OnPurchaseComplete 중간의 에러를 캐치?하면 OnPurchaseFailed가 실행된다는 생각을 못하고(별개로 호출되는 줄 알고), 제페토에서 제공된 IWP버튼부터의 과정에서 바로 OnPurchaseFailed가 실행된다고 생각해서,, 나는 틀릴 곳이 없다고 생각했네요🤗 에러 로그도 마침 안 떠서..

3. 틀렸던 부분은,, **함수 파라메터에 타입을 안 정해놨다가, string을 넣어야될 곳에 string[]을 넣었었습니다**

항상 모든 문제는 자신에게 있네여🤤 감사합니다

**Reply by Zepeto-TechEvangelist** [reply]
문제 해결에 도움이 되었다니 다행입니다 :)
내용 공유도 감사합니다.
